_posts/2020-03-30-method-s-arguments-pt-2.markdown

```diff
@@ -0,0 +1,350 @@
+
+```ruby
+def foo(**kargs) # it's common to name it kargs
+  puts "kargs is: #{kargs.inspect}"
+end
+```
+
+Now you can add as many keyword arguments as you want:
+
+```ruby
+foo(some_key: 1)
+# => kargs is: {:some_key=>1}
+
+foo(one_key:1, another_key: 2)
+# => kargs is: {:one_key=>1, :another_key=>2}
+```
+
+You can combine these arguments with required and optional arguments too:
+
+```ruby
+def foo(arg1:, arg2: 'default value', **kargs)
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+  puts "kargs is: #{kargs.inspect}"
+end
+
+foo(arg1: 1, arg2: 2, arg3: 3, arg4: 4)
+# => arg1 is: 1
+#    arg2 is: 2
+#    kargs is: {:arg3=>3, :arg4=>4}
+
+foo(arg1: 1, arg2: 2, arg3: 3)
+# => arg1 is: 1
+#    arg2 is: 2
+#    kargs is: {arg3: 3}
+
+foo(arg1: 1, arg2: 2)
+# => arg1 is: 1
+#    arg2 is: 2
+#    kargs is: {}
+
+foo(arg1: 1)
+# => arg1 is: 1
+#    arg2 is: default value
+#    kargs is: {}
+```
+
+Notice that "kargs" is always a hash. You usually deconstruct the hash (using the double splat operator) or use it as is as a hash or to call another method:
+
+```ruby
+def bar(arg2:, **kargs)
+  puts "arg2 in bar is: #{arg2.inspect}"
+  puts "kargs in bar is: #{kargs.inspect}"
+end
+
+def foo(arg1:, **kargs)
+  puts "arg1 in foo is: #{arg1.inspect}"
+  puts "kargs in foo is: #{kargs.inspect}"
+
+  # we need the ** (double splat operator) again so ruby deconstructs the
+  # hash into different keyword arguments
+  # if we don't do this it will show a
+  # deprecation warning in ruby 2.7 and
+  # raise an error in ruby 3
+  bar(**kargs)
+end
+
+foo(arg1: 1, arg2: 2, arg3: 3, arg4: 4)
+# => arg1 in foo is: 1
+#    kargs in foo is: {arg2: 2, arg3: 3, arg4: 4}
+#    arg2 in bar is: {arg2: 2}
+#    kargs in bar is: {arg3: 3, arg4: 4}
+```
+
+> This special parameter must always be used after the other keyword arguments. This is not allowed:
+  
+```ruby
+def foo(**kargs, arg1:)
+```
+
+## Combining positional and keyword arguments
+
+You can combine this two types of arguments but with the only requirement that positional arguments go first, then keyword arguments.
+
+```ruby
+def foo(arg1, arg2 = 'default for 2', *args, arg3:, arg4: 'default for 4', **kargs)
+```
+
+In that generic example we have:
+* arg1 => required by position
+* arg2 => optional by positional
+* args => optional by positional, variable
+* arg3 => required by key
+* arg4 => optional by key
+* kargs => optional by key, variable
+
+## Conclusion
+
+Now we can have methods with a lot of flexibility combining both positional and keyword arguments and making some of them required and some optionals, but we are not finished yet.
+
+Ruby gives us even more options that we will cover on the next post of this serie. We will also analyze a few common Rails methods to see how they use different types of arguments so we put everything together with practical examples.
```

Comment:
```suggestion
Ruby gives us even more options that we will cover on the next post in this series. We will also analyze a few common Rails methods to see how they use different types of arguments so that we can put everything together with practical examples.
```